src/connector/src/sink/iceberg/precomputed_partition_writer.rs

```diff
@@ -0,0 +1,282 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+
+use arrow_array::{BooleanArray, RecordBatch, StructArray};
+use arrow_row::{OwnedRow, RowConverter, SortField};
+use arrow_schema::{DataType, SchemaRef};
+use arrow_select::filter::filter_record_batch;
+use icelake::io_v2::{IcebergWriteResult, IcebergWriter, IcebergWriterBuilder};
+use icelake::types::{struct_to_anyvalue_array_with_type, Any, AnyValue};
+use risingwave_common::util::iter_util::ZipEqFast;
+
+#[derive(Clone)]
+pub struct PrecomputedPartitionedWriterBuilder<B: IcebergWriterBuilder> {
+    inner: B,
+    partition_type: Any,
+}
+
+impl<B: IcebergWriterBuilder> PrecomputedPartitionedWriterBuilder<B> {
+    pub fn new(inner: B, partition_type: Any) -> Self {
+        Self {
+            inner,
+            partition_type,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<B: IcebergWriterBuilder> IcebergWriterBuilder for PrecomputedPartitionedWriterBuilder<B> {
+    type R = PrecomputedPartitionedWriter<B>;
+
+    async fn build(self, schema: &SchemaRef) -> icelake::Result<Self::R> {
+        // Find the partition value column
+        let expect_partition_type =
+            self.partition_type.clone().try_into().map_err(|e| {
+                icelake::Error::new(icelake::ErrorKind::ArrowError, format!("{}", e))
+            })?;
+        let partition_col_idx = schema
```

Comment:
I think rather than looking this, it would be safer to pass the partition index to writer builder?